### PR TITLE
Homework 2 completed

### PR DESCRIPTION
**Pull request by aikido-sensei**

### CODE REVIEW

File: wiki/conditions/logical_operators.py [hunk 1]
```diff
@@ -57,6 +57,13 @@ def light_condition() -> bool:
 
 # TODO: Try the short circuit (|) & normal OR operators.
 
+#this "or" evaluates both condition even if the firs is true
```
Comment (the-c0d3br34k3r) [inline wiki/conditions/logical_operators.py]: *conditions, *first.

File: wiki/general/classes.py [hunk 2]
```diff
@@ -7,15 +7,15 @@ class Car:
     """
 
     def __init__(self) -> None:
-        self.__engine_on: bool = False
-        self.__speed_in_kmph: float = 0
+        self.engine_on: bool = False
+        self.speed_in_kmph: float = 0
```
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py]: Please revert the variables to private form.

File: wiki/general/classes.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    def brake_to_speed(self, desired_speed: float ) -> None :
+        """
+        Brake at a rate of 5kmph/s until the car is at the desired speed.
+        Desired_speed: Is the speed at which you want the car to be. 
+        """
+        if desired_speed > self.speed_in_kmph:
+            raise AttributeError("Cannot brake to a higher speed")
+        
+        if desired_speed < 0 :
+            desired_speed = 0
```
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py]: In such scenarios, throw an error.

File: wiki/general/classes.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    """
+    TODO: Add functionality to brake. Braking reduces the speed by 5kmph per second. Two brake methods. 
+    One to got to desired speed. Another to brake for specified duration.
+    """
+
+    def brake_to_speed(self, desired_speed: float ) -> None :
+        """
+        Brake at a rate of 5kmph/s until the car is at the desired speed.
+        Desired_speed: Is the speed at which you want the car to be.
```
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py]: Please fix documentation comments.

File: wiki/general/classes.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        while self.speed_in_kmph - desired_speed != 0:
+
+          if self.speed_in_kmph - desired_speed > 5:
+              self.speed_in_kmph -= 5
+              time.sleep(0.9999999999)
+        
+          else:
+              self.speed_in_kmph = self.speed_in_kmph - desired_speed
+              time.sleep(0.9999999)
+
```
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py]:
```suggestion
        time.sleep((self.speed_in_kmph - desired_speed)/5)
        self.speed_in_kmph = desired_speed
```

File: wiki/general/classes.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        for second in range(time): 
+
+          if self.speed_in_kmph >= 5:
+              self.speed_in_kmph -= 5
+              time.sleep(0.9999999999)
+        
+          else:
+              self.speed_in_kmph = 0
+              time.sleep(0.9999999)
+              break
```
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py]:
```suggestion
        time.sleep(time)
        self.speed_in_kmph = max(self.speed_in_kmph - time * 5, 0)
```

File: wiki/general/classes.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+          if self.speed_in_kmph - desired_speed > 5:
+              self.speed_in_kmph -= 5
+              time.sleep(0.9999999999)
+        
+          else:
+              self.speed_in_kmph = self.speed_in_kmph - desired_speed
+              time.sleep(0.9999999)
+
+
+    def brake_for(self, time: int ) -> None:
```
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py]:
```suggestion
    def brake_for(self, time_in_seconds: int ) -> None:
```

File: wiki/general/classes.py (hunk 6 above)
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py]:
```suggestion
        for second in range(time): 
          time.sleep(1)
          if self.speed_in_kmph > 1:
             self.speed_in_kmph = max(self.speed_in_kmph - 5, 0)
```

File: wiki/general/classes.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+print(car1.get_engine_on())
+print(car1.get_speed_in_kmph())
+
+car2: Car = Car()
+print(car2.get_engine_on())
+print(car2.get_speed_in_kmph())
+
+
+time.sleep(1)
+
```
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py:133]: Can you please add sample code for braking?

File: wiki/general/lambdas.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+square_function = lambda number: number * number
+sum_function = lambda number1, number2: number1 + number2
+is_even_function = lambda number: number % 2 == 0
+
+print(sum_function(3, 10))
+
+"""
+TODO: Try to define the functions as a part of the TODOs in the recursion.py file as lambda functions.
+"""
```
Comment (the-c0d3br34k3r) [inline wiki/general/lambdas.py:9]: TODO

File: wiki/loops/recursion.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    if n <= 1:
+        return accumulator
+    return factorial_using_tail_recursion(n - 1, n * accumulator)
+
+
+""" 
+    TODO: Try the above approaches for finding:
+     1. The sum on n numbers.
+     2. The value of pi - https://en.wikipedia.org/wiki/Leibniz_formula_for_π.
+"""
```
Comment (the-c0d3br34k3r) [inline wiki/loops/recursion.py:35]: TODO

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Please address the requested changes.

File: wiki/loops/recursion.py (hunk 10 above)
Comment (aikido-sensei) [reply wiki/loops/recursion.py:35]: Addressed in #3

File: wiki/general/lambdas.py (hunk 9 above)
Comment (aikido-sensei) [reply wiki/general/lambdas.py:9]: Addressed in #3

**Review by the-c0d3br34k3r — APPROVED** [review]

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Please rebase the branch with main.

File: wiki/general/classes.py [hunk 11]
```diff
@@ -7,8 +7,8 @@ class Car:
     """
 
     def __init__(self) -> None:
-        self.__engine_on: bool = False
-        self.__speed_in_kmph: float = 0
+        self.engine_on: bool = False
+        self.speed_in_kmph: float = 0
```
Comment (the-c0d3br34k3r) [inline wiki/general/classes.py]: Retain the variables as private.

**Review by the-c0d3br34k3r — CHANGES_REQUESTED** [review]
Requested changes and please rebase with the upstream main branch as well.

**Review by the-c0d3br34k3r — APPROVED** [review]

**Pull request merged**